[FLINK-7689] [Streaming Connectors] Added metrics to JDBCOutputFormat in order to be able to m…

asicoe opened the pull request:
Added metrics to JDBCOutputFormat in order to be able to measure JDBC batch flush rate, latency and size.

## What is the purpose of the change
This pull request adds some useful metrics to the flink-jdbc output format.

## Brief change log

- Added metrics to JDBCOutputFormat in order to be able to measure jdbc batch flush rate, latency and size.
- Fixed failing existing tests.
- Added tests for metrics setup.

## Verifying this change

This change added tests and can be verified as follows:
1. Run the JDBCOutputFormatTest suite.
2. Write a flink app that depends on flink-jdbc from this pr and flink-table and use it to write some records in a JDBC compliant database. Open the Flink UI and see the stats in the Task Metrics.
 
## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

fhueske commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCOutputFormat.java:
```diff
@@ -41,6 +46,11 @@
 	static final int DEFAULT_BATCH_INTERVAL = 5000;
+	static final String FLUSH_SCOPE = "flush";
+	static final String FLUSH_RATE_METER_NAME = "rate";
+	static final String FLUSH_RATE_GR_BATCH_INT_METER_NAME = "rateGreaterThanBatchInterval";
```
rename to `batchLimitReachedRate`?

fhueske commented on flink-connectors/flink-jdbc/src/test/java/org/apache/flink/api/java/io/jdbc/JDBCOutputFormatTest.java:
```diff
@@ -233,6 +255,30 @@ public void testFlush() throws SQLException, IOException {
 	}
 
+	@Test
+	public void testMetricsSetup() throws IOException {
```
Can you extend this test to check that the metrics are correctly set (except for the `durationMs` histogram)?

asicoe commented:
Thanks for reviewing Fabian. I have added the suggested changes. Please have a look. The build is failing for a non-related issue.

pabloem commented:
@fhueske @asicoe is this PR still current / ongoing? I'm willing to driving it to the end if there's anything left to do.. : ) - Or perhaps it's almost ready to merge?

zentol commented on flink-connectors/flink-jdbc/pom.xml:
```diff
@@ -59,5 +59,11 @@ under the License.
+
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-metrics-dropwizard</artifactId>
```
We should exclusively rely on built-in metrics.

zentol commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCOutputFormat.java:
```diff
@@ -82,6 +97,22 @@ public void open(int taskNumber, int numTasks) throws IOException {
+		this.flushDurationMsHisto = getRuntimeContext()
+			.getMetricGroup()
+			.addGroup(FLUSH_SCOPE)
+			.histogram(FLUSH_DURATION_HISTO_NAME, new DropwizardHistogramWrapper(new com.codahale.metrics.Histogram(new ExponentiallyDecayingReservoir())));
```
I recommend staying away form histograms as long as possible. Most metric backends recommend to _not_ build histograms in the application, but let the backend handle it.

zentol commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCOutputFormat.java:
```diff
@@ -82,6 +97,22 @@ public void open(int taskNumber, int numTasks) throws IOException {
+		this.flushMeter = getRuntimeContext()
+			.getMetricGroup()
+			.addGroup(FLUSH_SCOPE)
+			.meter(FLUSH_RATE_METER_NAME, new DropwizardMeterWrapper(new com.codahale.metrics.Meter()));
```
These could be replaced with the built-in `MeterView`.

zentol commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCOutputFormat.java:
```diff
@@ -58,6 +68,11 @@
 	private int[] typesArray;
 
+	private Meter batchLimitReachedMeter;
+	private Meter flushMeter;
```
These could be initialized in the constructor and made final.

asicoe replied:
they need the runtime to be initialised though

zentol commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCOutputFormat.java:
```diff
@@ -41,6 +46,11 @@
 	private static final long serialVersionUID = 1L;
 	static final int DEFAULT_BATCH_INTERVAL = 5000;
+	static final String FLUSH_SCOPE = "flush";
+	static final String FLUSH_RATE_METER_NAME = "rate";
```
I'm not convinced of the naming scheme. I would replace `FLUSH_SCOPE` with "jdbc", and explicitly prefix the rate and duration metrics with "flush".

zentol commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCOutputFormat.java:
```diff
@@ -207,13 +238,18 @@ public void writeRecord(Row row) throws IOException {
 
 		if (batchCount >= batchInterval) {
 			// execute batch
+			batchLimitReachedMeter.markEvent();
```
this seems redundant given that `flushMeter` exists. While the job is running `batchLimit == flushMeter`, and at the end `batchLimit == flushMeter -1` except in the exceedingly rare case that the total number of rows fits perfectly into the batches.

zentol commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCOutputFormat.java:
```diff
@@ -207,13 +238,18 @@ public void writeRecord(Row row) throws IOException {
+			flushBatchCountHisto.update(batchCount);
+			long before = System.currentTimeMillis();
 			upload.executeBatch();
+			flushDurationMsHisto.update(System.currentTimeMillis() - before);
```
This may result in a negative duration.

asicoe commented:
@zentol @pabloem @fhueske I've addressed all the latest comments, simplified the metrics (got rid of histograms) and updated tests. Please have a look!

dawidwys commented:
Hi @asicoe,
Sorry for the lack of activity in the PR. Let me know if you are still interested in contributing those metrics. I can help review this PR.

If not I will close this PR after some time.

dawidwys commented:
I'm closing this as "Abandoned", since there is no more activity and the code base has moved on quite a bit. Please re-open this if you feel otherwise and work should continue.
